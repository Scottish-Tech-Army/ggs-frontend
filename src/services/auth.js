const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function login(email) {
  return fetch(BASE_URL + "unit/login", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email }),
  }).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw response;
  })
}

export function register(email, name) {
  return fetch(BASE_URL + "unit/register", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ email, name }),
  }).then(response => {
    if (response.ok) {
      return response.json();
    }
    throw response;
  })
}
