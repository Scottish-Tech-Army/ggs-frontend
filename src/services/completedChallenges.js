const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;


// NOTE: in the original code the following function 
// was called getLeaderboard, hence the use of 
// "unit/leaderboard" in the first arg of fetch()
// below.
export function getCompletedChallenges(email) {
  return fetch(BASE_URL + "unit/leaderboard", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json()) // the .json() returns the actual data
    .catch((err) => {
      console.error(err);
    });
}
