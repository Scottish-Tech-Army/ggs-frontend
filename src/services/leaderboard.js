const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function getLeaderboard(email) {
  return fetch(BASE_URL + "unit/leaderboard", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json())
    .catch((err) => {
      console.error(err);
    });
}
