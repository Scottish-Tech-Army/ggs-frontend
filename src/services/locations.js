const BASE_URL = process.env.REACT_APP_AWS_CLIENT_API_ENDPOINT;

export function getLocations(email) {
  return fetch(BASE_URL + "locations", {
    method: "GET",
    headers: {
      GGSUnit: email,
    },
  })
    .then((data) => data.json())
    .catch((err) => {
      console.error(err);
    });
}

export function collectLocation(email, id) {
  return fetch(BASE_URL + "unit/collect", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      GGSUnit: email,
    },
    body: JSON.stringify({ id })
  }).catch((err) => {
    console.error(err);
  });
}
