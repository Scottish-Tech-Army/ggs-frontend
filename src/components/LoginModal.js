import React, { useRef, useState, useContext, useEffect } from "react";

// context:
import { authContext } from "../contexts/AuthContext";
import { login, newRegister } from "../services/auth";

// components:
import Button from "react-bootstrap/Button";
import GGSbuttonOne from "./GGSbuttonOne";
import Modal from "react-bootstrap/Modal";
import Form from "react-bootstrap/Form";

// images:
import blueDownArrow from '../assets/images/blueDownArrow.svg'



const LOGIN = "login";
const REGISTER = "register";


const LoginModal = ({ handleLoginClose, successOrFailResponse, whichModalScreen }) => {
  const { setUnit } = useContext(authContext);

  const [email, setEmail] = useState("");
  const [name, setName] = useState("");
  const [county, setCounty] = useState("");
  const [error, setError] = useState("");
  // const [state, setState] = useState(LOGIN);
  const [state, setState] = useState(whichModalScreen);

  useEffect(() => {
    setError("");
  }, [state]);


//--------------------------------------


//--------------------------------------

// Now a group of functions, etc to do with the dropdown menu:

// Two state properties to hold 
// i)  what the user types in
// ii) the filtered list of counties:
const [stateObj, setStateObj] = useState({
  typedText: '',
  filteredCounties: []
                                       })



// An array containing all of the counties of Scotland:
let allCountiesList = [
  {region: "Aberdeen and Shetland", key: "1"},
  {region: "Angus", key: "2"},
  {region: "Ayrshire North", key: "3"},
  {region: "Ayrshire South", key: "4"},
  {region: "Banff and Buchan", key: "5"},
  {region: "Dumfries and Galloway", key: "6"},
  {region: "Dunbartonshire", key: "7"},
  {region: "East Lothian", key: "8"},
  {region: "East Renfrewshire", key: "9"},
  {region: "Edinburgh", key: "10"},
  {region: "Fife", key: "11"},
  {region: "Forth Valley", key: "12"},
  {region: "Glasgow", key: "13"},
  {region: "Gordon", key: "14"},
  {region: "Highlands and Islands", key: "15"},
  {region: "Inverness-shire", key: "16"},
  {region: "Kincardine and Deeside", key: "17"},
  {region: "Midlothian", key: "18"},
  {region: "Moray", key: "19"},
  {region: "North Lanarkshire", key: "20"},
  {region: "Renfrewshire", key: "21"},
  {region: "Scottish Borders", key: "22"},
  {region: "South Lanarkshire", key: "23"},
  {region: "Tay and Ochils", key: "24"},
  {region: "West Lanarkshire", key: "25"},
  {region: "West Lothian", key: "26"}
                ]




// A ref to hold a boolean that code in the 
// return statement reads 
// to determine whether to show the dropdown 
// menu or not:
let showRegionsList = useRef(false)



// Now make array countiesToDisplay. Each of its
// members must be JSX for one menu item for the 
// dropdown menu. Each menu item is a div that 
// contains a <p> that is the name of one of the
// counties in state property array 
// filteredCounties, which is a filtered list 
// of counties:
const countiesToDisplay = 
    stateObj.filteredCounties.map((member)=>(
            <li key = {member.key} >
    <div className="regionsListItem" onClick = {()=> selectCountyFromDropDown(member.region) }>
    <p className="regionsListText">
    {member.region}     
    </p>
    </div>
    </li>
                                            )
                                 )


// The onClick handler of the divs that
// contain the county names in the 
// dropdown list.
// This fn must:
// 1) Put the county name into the input
// 2) Make the dropdown disappear
function selectCountyFromDropDown(countyName){
  // 1&2):
  makeListDisappear(countyName)
                                           }



// The onBlur handler for the search field.
// This makes the list of filtered/unfiltered
// region data disappear.
// This function must:
// 1) set stateObj.typedtext to the selected
//    county (to make the county name appear
//    in the input)
// 2) set showRegionsList.current to false
// to make the dropdown list disappear
// 3) Set state property county to the 
//    county selected from the dropdown menu:
function makeListDisappear(countyName){
  // 1):
  setStateObj(
      {
          typedText: countyName,
          filteredCounties: []   
      }
             )
  // 2):
  showRegionsList.current = false

  // 3) 
      setCounty(countyName)
                                      }
  


// The onChange handler for the input.
// This function constantly creates and updates 
// an array of all of the counties that contain the 
// letters that the user is typing in and puts that
// array into a state property.
// This function must:
//  1)   get what the user has typed in the input
//  2)   compare the typed-in text with the text that is  
//       each member of array allCountiesList. If the
//       typed-in text is in any member of allCountiesList 
//       save that member object to array newArray.
//  3)   set showRegionsList to true.
//  5)   Change state property typedText to what was typed in
//  6)   Set state property county to the typed in text
//       (this allows the form validation code in function 
//       handleRegistration to take action if the user forgot
//       to select a region from the dropdown list)
function displayFilteredRegionsList(e){
  let newArray = []
  // 1):
  let typedInText = e.target.value
  console.log(`Text typed in is ${typedInText}`)
  // 2): 
  // array allCountiesList has members like this: {region: "Aberdeenshire", key: "1"},
for (let i = 0; i < allCountiesList.length; i++) {
  let county = allCountiesList[i].region.toUpperCase();
  if (county.includes(typedInText.toUpperCase())) {
      newArray.push(allCountiesList[i])
                                                  } // end if
                                               } // end for
// 3):                                                
showRegionsList.current =true
// 5):
setStateObj({
  typedText: typedInText,
  filteredCounties: newArray 
      })

// 6):
setCounty(typedInText)  

                           } // end displayFilteredRegionsList


// The onFocus handler for the search field (the input of 
// className = "countyInput". This handler must:
// 1)   Put all the counties into 
//      stateObj.filteredCounties
function showAllRegionsList(e){
  // 1): 
  showRegionsList.current =true    
  setStateObj(
      {
          typedText: '',
          filteredCounties: allCountiesList   
      }   
             )
                            } // end showAllRegionsList

//--------------------------------------

/*
// NOTE: I couldn't get the validation error messages to show if 
// I used the following object as a state property. Maybe come 
// come back to this and work out why it failed to work!
let initStateErrorObject = {
showEmailErrorMess: false,
emailErrorMessage: " ",
showUnitNameErrorMess: false,
unitNameErrorMessage: " ",
showCountyErrorMess: false,
countyErrorMessage: " "
                      }

// Now a state property that is just like the ref above.
// function validate() sets this state property to the 
// ref above
const [errorObj, setErrorObj] = useState(initStateErrorObject)
*/



// Wed31May23 NOTE: I couldn't get the Register form validation error messages to
// show using an object in state to hold the show/hide error 
// flags and the error messages so I hard coded the error messages 
// and used these simple vars for the flags:
const [showRegisterEmailErrMess, setShowRegisterEmailErrMess] = useState(false)
const [showRegisterUnitNameErrMess, setShowRegisterUnitNameErrMess] = useState(false)
const [showRegisterCountyErrMess, setShowRegisterCountyErrMess] = useState(false)

// The same deal but for the Login form:
const [showLoginEmailErrMess, setShowLoginEmailErrMess] = useState(false)

//--------------------------------------

// Form input validation functions. One for the login form and
// one for the Registration form:

// A function to validate the single email input in
// the login form. This fn returns true when the 
// user has typed a valid email address and false when not:

function validateLoginInput(){
  let loginEmailIsValid = false  
  // Validate the email 
let validEmailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
if (email.match(validEmailRegex)) {
  console.log(`Log-in email is valid`)
  loginEmailIsValid = true;
  setShowLoginEmailErrMess(false)
                                  } else {
  console.log(`Log-in email is NOT valid`)
  setShowLoginEmailErrMess(true)
                                         }
return loginEmailIsValid
                             }


// A function to validate the user inputs for registration:
function validateRegistrationInputs (){
  
let emailIsValid = false  
let unitNameIsValid = false  
let countyIsValid = false 

// Validate the email 
let validEmailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/
if (email.match(validEmailRegex)) {
  // console.log(`email is valid`)
  emailIsValid = true;
  setShowRegisterEmailErrMess(false)
                                  } else {
  // console.log(`email is NOT valid`)
  setShowRegisterEmailErrMess(true)
                                         }


// Validate the unit name:
if (name.trim().length === 0 || name === "") {
  setShowRegisterUnitNameErrMess(true)
                                             } else {
  unitNameIsValid = true
  setShowRegisterUnitNameErrMess(false)
                                                    }


// Validate the county name
for (const memberObj of allCountiesList){
  if (memberObj.region === county) {
    countyIsValid = true
    break
                                   }
                                        } // end for
if (!countyIsValid) {
  setShowRegisterCountyErrMess(true)
                    } else {
    setShowRegisterCountyErrMess(false)
                           }


if (emailIsValid && unitNameIsValid && countyIsValid) {
    return true // all fields passed validation
                                                      } else {
    return false // validation failed somewhere
                                                             }
                                } // end validateRegistrationInputs



//--------------------------------------
// The functions that handle login and registration:

  const handleLogin = (event) => {
    event.preventDefault(); // This line stops the modal window from
                            // vanishing even when the email address is invalid.
    // Only send the form data (an email address) to the back end 
    // if it passes validation:
    if (validateLoginInput()) {
      // Log in:
    login(email)   // email here is a JS object like this: {email: "xxx.yyy@zzz.com”}
      .then((unit) => {
        // login() returns a promise that resolves to unit,
        // which is a js object that looks like this:
        // {email: "mukund.pandit@virginmedia.com",
        // name: "TeamMukund"}

        setUnit(unit);
        handleLoginClose();
        // Call parent <LandingPage/>'s function that makes the app's
        // buttons operable and opaque (before login they are faded and 
        // inoperable).:
        successOrFailResponse(true)
      })
      .catch((error) => {
        // The function that tells parent <LandingPage/>
        // what to do on failure to log in:
        successOrFailResponse(false)
        // console.error(error);
        if (error.status === 404) {
          setError(
            "Email address not found. If this is your first visit, please register first."
          );
        } else {
          setError("Problem logging in. Please try again.");
        }
      });
    event.preventDefault();
                          } // end if
                            }; // end handleLogin()



// The event handler for the Register form's 
// Register button
  const handleRegistration = (event) => {
    // Validate user input:
    if (validateRegistrationInputs()) {
// console.log(`All inputs passed validation!`)
/*  NEW CODE FOLLOWS      
    newRegister(email, name, county)
      .then((unit) => {
        setUnit(unit);
        handleLoginClose();
      })
      .catch((error) => {
        console.error(error);
        if (error.status === 409) {
          setError("Email address already registered. Please log in instead.");
        } else {
          setError("Problem registering. Please try again.");
        }
      });
                                } // end if
*/                                

    /* OLD CODE FOLLOWS
    register(email, name)
      .then((unit) => {
        setUnit(unit);
        handleLoginClose();
      })
      .catch((error) => {
        console.error(error);
        if (error.status === 409) {
          setError("Email address already registered. Please log in instead.");
        } else {
          setError("Problem registering. Please try again.");
        }

      });
    */
                                } // end if
    event.preventDefault();
  };

  return (
    <Modal
      show={true}
      onHide={handleLoginClose}
      backdrop="static"
      keyboard={false}
      className="custom-modal login-modal"
    >
      <Modal.Header className="border-0 mb-n3"></Modal.Header>
      <Modal.Body className="mt-n3">
        {state === LOGIN && (
          <>
            <h1 style={{ textAlign: "center", marginBottom: "50px" }}>Ready to explore?</h1>
            {/*  tues30May23: The following few lines were 
              the original text, which seemed unnecessary to me, so I have removed them. Mukund
            <p className="text-left mb-3">
              Visit and collect all of the sights in your region or explore
              others
            </p>
            */}
            {error && <p className="error-text text-center mt-n3">{error}</p>} {/* The messsage that shows when there's a login error */}
            <Form noValidate className="m-2 mt-n2" onSubmit={handleLogin}>
              <Form.Group controlId="formBasicCode">
                {/* if user typed an invalid email, the following error message shows: */}
              {showLoginEmailErrMess &&  
              <p style={{color: "red", fontStyle: "italic"}} className="registerErrorMessage">Please type a valid email address</p>
              }
                <Form.Control
                  value={email}
                  type="email"
                  onChange={(e) => setEmail(e.target.value)}
                  label="email address"
                  placeholder="Enter email address"
                  className="w-100 my-2 mx-auto"
                />
              </Form.Group>
              <div className="modalButtonContainer">
              <Button 
              variant = "customGGS"
              className="buttonOperable" 
              type="submit"
              >
              Start Exploring
              </Button>
              <GGSbuttonOne
              buttonDivCSSclass = {"largeButton1New"}
              pTextCSSclass = {"buttonOperable"}
              clickHandler = {handleLoginClose}
              pText = "Cancel"
              />
              </div>
              </Form>
            <p className="text-left mt-3">
              Is this your first visit? Please{" "}
              <span className="switch-mode" onClick={() => setState(REGISTER)}>
                register first
              </span>
              .
            </p>
          </>
        )}
        {state === REGISTER && (
          <>
            <h1 style={{ textAlign: "center" }}>How To Register</h1>
            <div className="registerModalGrid">
            <div className="numberCircle numbCircleDark"><p className="numberCircleText">1</p></div>
            <div>
            <p className="mb-3">
              First type an email address, preferably your unit's shared address.
              We'll hold the address according to our {" "}
              <a
                href="https://www.girlguidingscotland.org.uk/privacy/"
                target="_blank"
                rel="noreferrer"
              >
                privacy policy
              </a>
              .
              </p>
            </div>

            <div className="numberCircle numbCircleDark"><p className="numberCircleText">2</p></div> 
            <div>
            <p className="mb-3">
              Then type your team's name.
              </p>
            </div>

            <div className="numberCircle numbCircleDark"><p className="numberCircleText">3</p></div> 
            <div>
            <p className="mb-3">
              Then select your county.
              </p>
            </div>

            <div className="numberCircle numbCircleDark"><p className="numberCircleText">4</p></div> 
            <div>
            <p className="mb-3">
              Then click Register or Cancel.
              </p>
            </div>
           
            </div>
              
            <div className="downArrowContainer">
              <img className="blueDownArrowIMG" src={blueDownArrow}></img>
            </div>  
            
            {error && <p className="error-text text-center mt-n3">{error}</p>} {/* The error that shows when the email is 
                                                                                already registered or there was a problem
                                                                                at the backend */} 
            <Form noValidate className="m-2 mt-n2" onSubmit={handleRegistration}>
              <Form.Group controlId="formBasicCode">
              <div className="numberCircle numbCircleLight"><p className="numberCircleText">1</p></div>
              {/* if user typed invalid email, this error message shows */}
              {showRegisterEmailErrMess &&  
              <p style={{color: "red", fontStyle: "italic"}} className="registerErrorMessage">Please type a valid email address</p>
              }
                <Form.Control
                  
                  value={email}
                  type="email"
                  onChange={(e) => {console.log(`The email address is: ${e.target.value}`)
                    setEmail(e.target.value)}}
                  label="email address"
                  placeholder="Type email address"
                  className="w-100 my-2 mx-auto formInputs" /* NOTE: I added class formInputs -- mukund*/
                />
                <div className="numberCircle numbCircleLight"><p className="numberCircleText">2</p></div>
                {/* if user typed invalid unit name, this error message shows */}
                {showRegisterUnitNameErrMess && 
              <p style={{color: "red", fontStyle: "italic"}} className="registerErrorMessage">Please type a unit name below</p>
              }
                <Form.Control
                  value={name}
                  onChange={(e) => setName(e.target.value)}
                  label="team name"
                  placeholder="Type team name"
                  className="w-100 my-2 mx-auto formInputs"  /* NOTE: I added class formInputs -- mukund*/
                />
                <div className="numberCircle numbCircleLight"><p className="numberCircleText">3</p></div>
                {/* if user typed invalid county name, this error message shows */}
                {showRegisterCountyErrMess && 
              <p style={{color: "red", fontStyle: "italic"}} className="registerErrorMessage">Please start typing a county name then select an option from the dropdown menu that appears</p>
                }
                <Form.Control
                  value={stateObj.typedText} // use of value makes this what React calls a controlled input
                  // value={county} // previous code
                  onChange={displayFilteredRegionsList}
                  // onChange={(e) => setCounty(e.target.value)} // previous code
                  onFocus = {showAllRegionsList} // user clicks in input and all counties show in the dropdown list
                  label="county name"
                  placeholder="Type county name then select from dropdown menu"
                  className="w-100 my-2 mx-auto formInputs"  /* NOTE: I added class formInputs -- mukund*/
                />
              </Form.Group>
              {/* Conditionally show/hide the dropdown list */}
              <div className="positionDDlistInLoginModal"> {/* This div contains the DD menu. 
              Its height is 5px, has overflow-y set to visible and not visible.
              The idea is to have the dd menu appear without forcong the other 
              elements in the <Form/> out of the way*/}
              <div className= {showRegionsList.current ? " showDDlist " : " hideDDlist"}>  
              <ul className="countiesList">
              {countiesToDisplay}
              </ul>
              </div>
              </div>
  
              <div className="numberCircle numbCircleLight"><p className="numberCircleText">4</p></div>
              <div className="modalButtonContainer">
              <Button 
              variant = "customGGS"
              className="buttonOperable" 
              type="submit">
                Register
              </Button>
              <GGSbuttonOne
              buttonDivCSSclass = {"largeButton1New"}
              pTextCSSclass = {"buttonOperable"}
              clickHandler = {handleLoginClose}
              pText = "Cancel"
              />
              </div>
              <div className="positionFinalText">
            {/*} <p className="text-center mt-3" >*/}
            <p className="finalText">
              Already registered? Please{" "}
              <span className="switch-mode" onClick={() => setState(LOGIN)}>
                log in
              </span>
              .
            </p>
            </div>
            </Form>
            
          </>
        )}
        </Modal.Body>
    </Modal>
  );
};

export default LoginModal;
