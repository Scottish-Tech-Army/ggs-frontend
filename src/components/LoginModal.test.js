/* eslint-disable testing-library/no-node-access */
import React from "react";
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginModal from "./LoginModal";
import { login, register } from "../services/auth";
import AuthProvider from "../contexts/AuthContext";

const handleLoginClose = jest.fn();

jest.mock("../services/auth");
const localStorageGetItem = jest.spyOn(
  Object.getPrototypeOf(window.localStorage),
  "getItem"
);
const localStorageSetItem = jest.spyOn(
  Object.getPrototypeOf(window.localStorage),
  "setItem"
);

describe("component LoginModal", () => {
  const TEST_EMAIL = "test@example.com";
  const TEST_TEAM_NAME = "Test team";
  const TEST_UNIT = { email: TEST_EMAIL, name: TEST_TEAM_NAME };

  beforeEach(() => {
    localStorageGetItem.mockReturnValue(null);
    localStorageSetItem.mockReturnValue(null);
  });

  afterAll(() => {
    localStorageGetItem.mockRestore();
    localStorageSetItem.mockRestore();
  });

  function renderWithUser() {
    return {
      user: userEvent.setup(),
      ...render(
        <AuthProvider>
          <LoginModal handleLoginClose={handleLoginClose} />
        </AuthProvider>
      ),
    };
  }

  it("switch between views", async () => {
    const { user } = renderWithUser();

    const modal = document.querySelector(".modal-dialog");
    expect(screen.getByText("Ready to explore?")).toBeDefined();
    expect(modal).toMatchSnapshot();

    await user.click(screen.getByText("register first"));

    expect(screen.getByText("Register Team")).toBeDefined();
    expect(modal).toMatchSnapshot();

    await user.click(screen.getByText("log in"));

    expect(screen.getByText("Ready to explore?")).toBeDefined();
  });

  it("login success", async () => {
    const { user } = renderWithUser();
    login.mockResolvedValue(TEST_UNIT);

    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByText("Start Exploring"));
    });

    await waitFor(() => expect(handleLoginClose).toHaveBeenCalledTimes(1));

    expect(login).toHaveBeenCalledTimes(1);
    expect(login).toHaveBeenCalledWith(TEST_EMAIL);

    expect(localStorageSetItem).toHaveBeenCalledTimes(1);
    expect(localStorageSetItem).toHaveBeenCalledWith(
      "ggsUnit",
      JSON.stringify(TEST_UNIT)
    );
  });

  it("login failure - unknown email", async () => {
    const { user } = renderWithUser();
    login.mockRejectedValue({ status: 404 });

    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByText("Start Exploring"));
    });

    expect(
      await screen.findByText(
        "Email address not found. If this is your first visit, please register first."
      )
    ).toBeDefined();

    expect(login).toHaveBeenCalledTimes(1);
    expect(login).toHaveBeenCalledWith(TEST_EMAIL);

    expect(localStorageSetItem).not.toHaveBeenCalled();
    expect(handleLoginClose).not.toHaveBeenCalled();
  });

  it("login failure - other error", async () => {
    const { user } = renderWithUser();
    login.mockRejectedValue({ status: 500 });

    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByText("Start Exploring"));
    });

    expect(
      await screen.findByText("Problem logging in. Please try again.")
    ).toBeDefined();

    expect(login).toHaveBeenCalledTimes(1);
    expect(login).toHaveBeenCalledWith(TEST_EMAIL);

    expect(localStorageSetItem).not.toHaveBeenCalled();
    expect(handleLoginClose).not.toHaveBeenCalled();
  });

  it("register success with team name", async () => {
    const { user } = renderWithUser();
    register.mockResolvedValue(TEST_UNIT);

    await user.click(screen.getByText("register first"));
    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    await user.type(
      screen.getByPlaceholderText("Enter team name (optional)"),
      TEST_TEAM_NAME
    );
    act(() => {
      user.click(screen.getByRole("button", { name: "Register" }));
    });

    await waitFor(() => expect(handleLoginClose).toHaveBeenCalledTimes(1));

    expect(register).toHaveBeenCalledTimes(1);
    expect(register).toHaveBeenCalledWith(TEST_EMAIL, TEST_TEAM_NAME);

    expect(localStorageSetItem).toHaveBeenCalledTimes(1);
    expect(localStorageSetItem).toHaveBeenCalledWith(
      "ggsUnit",
      JSON.stringify(TEST_UNIT)
    );
  });

  it("register success without team name", async () => {
    const { user } = renderWithUser();
    register.mockResolvedValue({ email: TEST_EMAIL });

    await user.click(screen.getByText("register first"));
    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByRole("button", { name: "Register" }));
    });

    await waitFor(() => expect(handleLoginClose).toHaveBeenCalledTimes(1));

    expect(register).toHaveBeenCalledTimes(1);
    expect(register).toHaveBeenCalledWith(TEST_EMAIL, "");

    expect(localStorageSetItem).toHaveBeenCalledTimes(1);
    expect(localStorageSetItem).toHaveBeenCalledWith(
      "ggsUnit",
      JSON.stringify({ email: TEST_EMAIL })
    );
  });

  it("register failure - existing email", async () => {
    const { user } = renderWithUser();
    register.mockRejectedValue({ status: 409 });

    await user.click(screen.getByText("register first"));
    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByRole("button", { name: "Register" }));
    });

    expect(
      await screen.findByText(
        "Email address already registered. Please log in instead."
      )
    ).toBeDefined();

    expect(register).toHaveBeenCalledTimes(1);
    expect(register).toHaveBeenCalledWith(TEST_EMAIL, "");

    expect(localStorageSetItem).not.toHaveBeenCalled();
    expect(handleLoginClose).not.toHaveBeenCalled();
  });

  it("register failure - other error", async () => {
    const { user } = renderWithUser();
    register.mockRejectedValue({ status: 500 });

    await user.click(screen.getByText("register first"));
    await user.type(
      screen.getByPlaceholderText("Enter email address"),
      TEST_EMAIL
    );
    act(() => {
      user.click(screen.getByRole("button", { name: "Register" }));
    });

    expect(
      await screen.findByText("Problem registering. Please try again.")
    ).toBeDefined();

    expect(register).toHaveBeenCalledTimes(1);
    expect(register).toHaveBeenCalledWith(TEST_EMAIL, "");

    expect(localStorageSetItem).not.toHaveBeenCalled();
    expect(handleLoginClose).not.toHaveBeenCalled();
  });
});
