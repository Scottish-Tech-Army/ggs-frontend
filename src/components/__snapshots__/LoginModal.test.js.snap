// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component LoginModal switch between views 1`] = `
<div
  class="modal-dialog"
>
  <div
    class="modal-content"
  >
    <div
      class="border-0 mb-n3 modal-header"
    />
    <div
      class="mt-n3 modal-body"
    >
      <h1
        style="text-align: center;"
      >
        Ready to explore?
      </h1>
      <p
        class="text-center mb-3"
      >
        Visit and collect all of the sights in your region or explore others
      </p>
      
      <form
        class="m-2 mt-n2"
      >
        <div
          class="form-group"
        >
          <input
            class="w-100 my-2 mx-auto form-control"
            id="formBasicCode"
            label="email address"
            placeholder="Enter email address"
            type="email"
            value=""
          />
        </div>
        <button
          class="btn-branding w-100 my-2 mx-auto btn-branding w-100 my-2 mx-auto-primary"
          type="submit"
        >
          Start Exploring
        </button>
      </form>
      <p
        class="text-center mt-3"
      >
        Is this your first visit? Please
         
        <span
          class="switch-mode"
        >
          register first
        </span>
        .
      </p>
    </div>
  </div>
</div>
`;

exports[`component LoginModal switch between views 2`] = `
<div
  class="modal-dialog"
>
  <div
    class="modal-content"
  >
    <div
      class="border-0 mb-n3 modal-header"
    />
    <div
      class="mt-n3 modal-body"
    >
      <h1
        style="text-align: center;"
      >
        Register Team
      </h1>
      <p
        class="text-center mb-3"
      >
        Enter an email and optional name to register your team. We will hold your email address in accordance with our
         
        <a
          href="https://www.girlguidingscotland.org.uk/privacy/"
          rel="noreferrer"
          target="_blank"
        >
          privacy policy
        </a>
        .
      </p>
      
      <form
        class="m-2 mt-n2"
      >
        <div
          class="form-group"
        >
          <input
            class="w-100 my-2 mx-auto form-control"
            id="formBasicCode"
            label="email address"
            placeholder="Enter email address"
            type="email"
            value=""
          />
          <input
            class="w-100 my-2 mx-auto form-control"
            id="formBasicCode"
            label="team name"
            placeholder="Enter team name (optional)"
            value=""
          />
        </div>
        <button
          class="btn-branding w-100 my-2 mx-auto btn-branding w-100 my-2 mx-auto-primary"
          type="submit"
        >
          Register
        </button>
      </form>
      <p
        class="text-center mt-3"
      >
        Already registered? Please
         
        <span
          class="switch-mode"
        >
          log in
        </span>
        .
      </p>
    </div>
  </div>
</div>
`;
